File: C_Programming/Practise_Programs/PP_137.c
//////////////////////////////////////////////////////////////////
//
//  File Name         : PP_137
//
//  Description       : This program demonstrates string reversal
//                      using two character pointers and swapping
//                      characters in the string.
//
//  Author            : Shubham Somanath Gadhe
//  Date              : 17/08/2026
//
//  Time Complexity   : O(n)
//  Space Complexity  : O(n)
//
//////////////////////////////////////////////////////////////////

#include<stdio.h>

//////////////////////////////////////////////////////////////////
//
//  Function Name     : strrevX()
//
//  Description       : It is used to reverse the given string by
//                      swapping characters using two character
//                      pointers.
//
//  Author            : Shubham Somanath Gadhe
//  Date              : 17/08/2026
//
//////////////////////////////////////////////////////////////////

void strrevX(char *str)
{
    char *start  = NULL;
    char *end = NULL;
    char temp = '\0';
    
    start = str;

    while(*str != '\0')
    {
        str++;
    }

    str--;

    end = str;
    
    while(start < end)
    {
        temp = *start;
        *start = *end;
        *end = temp;
        
        start++;
        end--;
    }
}

int main()
{
    char Arr[50] = {'\0'};

    printf("Enter String : \n");
    scanf("%[^'\n]s", Arr);

    strrevX(Arr);

    printf("Updated string : %s\n", Arr);

    return 0;
}

//////////////////////////////////////////////////////////////////
//
//  Sample Input :
//      Ganesh
//
//  Sample Output :
//
//      Enter String :
//      Updated string : hsenaG
//
//////////////////////////////////////////////////////////////////

File: Java_Programming/Practise_Programs/PP_137.java
//////////////////////////////////////////////////////////////////
//
//  File Name         : PP_137
//
//  Description       : This program prints a butterfly star pattern
//                      using nested for loops. The pattern consists
//                      of two symmetrical halves with stars and
//                      spaces.
//
//  Author            : Shubham Somanath Gadhe
//  Date              : 31/08/2026
//
//  Time Complexity   : O(n^2)
//  Space Complexity  : O(1)
//
//////////////////////////////////////////////////////////////////

class PP_137
{
    public static void main(String args[])
    {
        for(int i = 1; i <= 5; i++)              
        {
            for(int j = 1; j <= i; j++)          
            {
                System.out.print("*");
            }

            for(int j = 1; j <= (5 - i) * 2; j++) 
            {
                System.out.print(" ");
            }

            for(int j = 1; j <= i; j++)          
            {
                System.out.print("*");
            }

            System.out.println();
        }

        for(int i = 1; i <= 4; i++)              
        {
            for(int j = 1; j <= 5 - i; j++)      
            {
                System.out.print("*");
            }

            for(int j = 1; j <= i * 2; j++)      
            {
                System.out.print(" ");
            }

            for(int j = 1; j <= 5 - i; j++)      
            {
                System.out.print("*");
            }

            System.out.println();
        }
    }
}

//////////////////////////////////////////////////////////////////
//
//  Sample Input :
//      No input required.
//
//  Sample Output :
//
//      *        *
//      **      **
//      ***    ***
//      ****  ****
//      **********
//      ****  ****
//      ***    ***
//      **      **
//      *        *
//
//////////////////////////////////////////////////////////////////